Arrowgene.Ddon.GameServer/Handler/DispelExchangeDispelItemHandler.cs

```diff
@@ -0,0 +1,153 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Arrowgene.Ddon.GameServer.Handler
+{
+    public class DispelExchangeDispelItemHandler : GameRequestPacketHandler<C2SDispelExchangeDispelItemReq, S2CDispelExchangeDispelItemRes>
+    {
+        private static readonly ServerLogger Logger = LogProvider.Logger<ServerLogger>(typeof(DispelExchangeDispelItemHandler));
+
+        public DispelExchangeDispelItemHandler(DdonGameServer server) : base(server)
+        {
+        }
+
+        public override S2CDispelExchangeDispelItemRes Handle(GameClient client, C2SDispelExchangeDispelItemReq request)
+        {
+            var res = new S2CDispelExchangeDispelItemRes();
+
+            var appraisialItems = Server.AssetRepository.SpecialShopAsset.AppraisalItems;
+            foreach (var item in request.GetDispelItemList)
+            {
+                if (!appraisialItems.ContainsKey(item.Id))
+                {
+                    throw new ResponseErrorException(ErrorCode.ERROR_CODE_ITEM_NOT_FOUND);
+                }
+
+                bool toBag = false;
+                switch (item.StorageType)
+                {
+                    case 19:
+                        toBag = true;
+                        break;
+                    case 20:
+                        toBag = false;
+                        break;
+                    default:
+                        throw new ResponseErrorException(ErrorCode.ERROR_CODE_ITEM_INVALID_STORAGE_TYPE, $"Unexpected destination when exchanging items {item.StorageType}");
+                }
+
+                // Check for cost
+                bool hasFullPayment = true;
+                foreach (var payment in item.UIDList)
+                {
+                    hasFullPayment &= Server.ItemManager.HasItem(Server, client.Character, ItemManager.BothStorageTypes, payment.UId, payment.Num);
+                }
+
+                if (!hasFullPayment)
+                {
+                    throw new ResponseErrorException(ErrorCode.ERROR_CODE_DISPEL_LACK_MONEY);
+                }
+
+                var updateCharacterItemNtc = new S2CItemUpdateCharacterItemNtc()
+                {
+                    UpdateType = ItemNoticeType.GetDispelItem
+                };
+
+                // Consume payment
+                foreach (var payment in item.UIDList)
+                {
+                    List<CDataItemUpdateResult> updateResults = Server.ItemManager.ConsumeItemByUIdFromMultipleStorages(Server, client.Character, ItemManager.BothStorageTypes, payment.UId, payment.Num);
+                    updateCharacterItemNtc.UpdateItemList.AddRange(updateResults);
+                }
+
+                var purchase = AppraiseItem(client.Character, appraisialItems[item.Id]);
+
+                List<CDataItemUpdateResult> itemUpdateResults = Server.ItemManager.AddItem(Server, client.Character, toBag, purchase.ItemId, purchase.ItemNum);
+                Debug.Assert(itemUpdateResults.Count == 1);
```

Comment:
Debug.Assert doesn't run when the project is compiled in Release mode

Reply:
Replaced it with an exception, although I couldn't find a very good error code for it.